ROADMAP.md

```diff
@@ -0,0 +1,579 @@
+identity, etc. may simplify things. An open question is where
+Multistream would fit into this model: could we conflate it into the
+Wire protocol?
+
+**Why?** Peers act haphazardly when observed from the outside, as they
+do not have the means to act more collaboratively. We are lacking a wire
+protocol that governs the connection between two peers. Negotiating lots
+of piecemeal protocols is inefficient, especially for essential stuff.
+
+**Links:**
+
+-   [Discuss: batch vs. on-demand connection
+    pruning](https://github.com/libp2p/go-libp2p-connmgr/issues/19).
+
+
+### ☎️ Reducing the dial fail rate
+
+**What?** Reducing the dial fail rate implies improvements in several
+areas.
+
+-  **Outbound dials:** We should strategise dials based on context, or based on
+   a heuristic determined by the application. For example, apps designed for
+   local exchange of data will naturally prefer to try local/physical endpoints
+   over public ones first. We may want to dial relay addresses last, dial newer
+   addresses first, etc.
+
+   - Users should also have the capability to customise how addresses
+     are resolved, how dials are throttled, how connections are
+     selected if multiple dials succeeded, etc. In a nutshell, the
+     dialing behaviour should be fully customisable per host.
+
+-  **Context-sensitive advertisement of self-addresses:** private
+   addresses should not leak into public DHTs, both for connectivity
+   and security reasons (cf. IPFS DHT pollution in 2016). Similarly,
+   local discovery mechanisms should only transmit private addresses.
+   The DHT topology question becomes relevant to define scopes (e.g.
+   federated/concentric/shared DHTs for different scopes). Also, when
+   dealing with multi-layer NATs, there may be better routes between
+   peers behind regional or ISP NATs other than their public
+   endpoints.
+
+- **Connectivity-related problems:** see [Unprecedented global
+  connectivity](#unprecedented-global-connectivity) and
+  [Improve NAT traversal](#improve-nat-traversal).
+
+**Why?** Dialing is a crucial function of our system and using a
+non-deterministic, scattershot approach is inefficient.
+
+**Links:**
+
+- [Reliable dialing](https://github.com/libp2p/libp2p/issues/27).
+
+### 🔀 Peer exchange protocol
+
+Copied and adapted from:
+[https://github.com/libp2p/notes/issues/3](https://github.com/libp2p/notes/issues/3)
+
+**What?** We currently use mDNS to discover local peers but it would
+also be nice if we could have a peer exchange protocol for discovering
+even more local peers.
+
+**Why?** (1) relying on the global DHT to discover local peers is
+problematic; (2) ideally, we want to connect to all peers on the local
+network for UX reasons (if we\'re in the same room, we should be able to
+share files with each-other). [This ties into offline and decentralised
+networks.](#kix.xlvu7ikvo1jg)
+
+### 🏹 RPC and other common node communication patterns
```

Comment:
This one seems interesting, particularly from it's ability to unlock others.  

I think it could benefit from something like:

```
This is what a customer (developer) has to do today to satisfy usecase Y.
...List the various steps/commands...

If we do this work, they will instead only need to do:
...List the various steps/commands...
```

Reply:
For now, I would prefer to keep this section on a high level, especially as each of the detailed steps needed by users today is very implementation specific.